File: error-notebook/scripts/review-notebook.mjs
#!/usr/bin/env node
/** 扫描待复核、已经超过复核日期及生命周期字段异常的错误经验。 */

import fs from "node:fs";
import path from "node:path";

import {
  extractLifecycle,
  invalidLifecycleLabel,
  lifecycleLabel,
  parseIsoDate,
  todayDate,
} from "./notebook-lifecycle.mjs";
import { parseNotebook } from "./notebook-parser.mjs";
import { activeNotebookPath } from "./notebook-paths.mjs";


/** 解析复核扫描参数。 */
function parseArguments(argumentsList) {
  let notebook;
  let asOf = todayDate();
  let maxResults = 20;
  let failOnAction = false;

  for (let index = 0; index < argumentsList.length; index += 1) {
    const argument = argumentsList[index];
    if (argument === "--notebook") {
      index += 1;
      if (index >= argumentsList.length) {
        throw new Error("--notebook 缺少路径。");
      }
      notebook = path.resolve(argumentsList[index]);
    } else if (argument === "--as-of") {
      index += 1;
      if (index >= argumentsList.length) {
        throw new Error("--as-of 缺少日期。");
      }
      asOf = parseIsoDate(argumentsList[index]);
      if (!asOf) {
        throw new Error("日期必须使用 YYYY-MM-DD 格式。");
      }
    } else if (argument === "--max-results") {
      index += 1;
      if (index >= argumentsList.length) {
        throw new Error("--max-results 缺少数量。");
      }
      maxResults = Number(argumentsList[index]);
    } else if (argument === "--fail-on-action") {
      failOnAction = true;
    } else {
      throw new Error(`未知参数：${argument}`);
    }
  }

  if (!Number.isInteger(maxResults) || maxResults < 1) {
    throw new Error("--max-results 必须是大于 0 的整数。");
  }
  notebook ??= activeNotebookPath();
  return { notebook, asOf, maxResults, failOnAction };
}

/** 输出生命周期汇总和需要处理的条目。 */
function main() {
  let options;
  try {
    options = parseArguments(process.argv.slice(2));
  } catch (error) {
    console.error(`错误：${error.message}`);
    return 2;
  }
  if (!fs.existsSync(options.notebook) || !fs.statSync(options.notebook).isFile()) {
    console.error(`错误：错题本不存在：${options.notebook}`);
    return 2;
  }

  let text;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(
      fs.readFileSync(options.notebook),
    );
  } catch (error) {
    console.error(`错误：错题本不是有效 UTF-8：${error.message}`);
    return 2;
  }
  const { entries, hasUnclosedFence } = parseNotebook(text);
  if (hasUnclosedFence) {
    console.error("错误：存在未闭合的 Markdown 围栏代码块，无法可靠扫描。");
    return 2;
  }
  const labels = entries.map((entry) => ({
    entry,
    label: lifecycleLabel(entry.lines, options.asOf),
  }));
  const counts = new Map();
  for (const { label } of labels) {
    counts.set(label, (counts.get(label) ?? 0) + 1);
  }
  const actionable = labels.filter(
    ({ label }) => label === "待复核" || label === "已到期" || label === invalidLifecycleLabel,
  );

  console.log(`# 错题本复核扫描（截至 ${options.asOf.toISOString().slice(0, 10)}）`);
  console.log();
  console.log(`- 总条目：${entries.length}`);
  for (const label of ["有效", "待复核", "已到期", "已失效", "已替代", invalidLifecycleLabel]) {
    console.log(`- ${label}：${counts.get(label) ?? 0}`);
  }

  console.log("\n## 待处理条目\n");
  if (actionable.length === 0) {
    console.log("没有待复核、已到期或字段异常条目。");
    return 0;
  }

  for (const { label, entry } of actionable.slice(0, options.maxResults)) {
    const values = extractLifecycle(entry.lines);
    console.log(
      `- [${label}] ${entry.category} / ${entry.title}；`
      + `最后验证：${values.lastVerified}；下次复核：${values.reviewBy}`,
    );
  }
  if (counts.has(invalidLifecycleLabel)) {
    console.log("字段异常条目请使用 validate-notebook.mjs 校验同一数据文件，定位具体原因。");
  }
  const remaining = actionable.length - options.maxResults;
  if (remaining > 0) {
    console.log(`- 其余 ${remaining} 条未展开；使用 --max-results 调整显示数量。`);
  }
  return options.failOnAction ? 1 : 0;
}


process.exitCode = main();

File: error-notebook/scripts/review_notebook.py
#!/usr/bin/env python3
"""扫描待复核和已经超过复核日期的错误经验。"""

from __future__ import annotations

import argparse
from collections import Counter
from datetime import date
from pathlib import Path
import sys

from notebook_lifecycle import extract_lifecycle, lifecycle_label
from search_notebook import DEFAULT_NOTEBOOK, parse_notebook


def parse_as_of(value: str) -> date:
    """把命令行日期解析为 ISO 日期。"""

    try:
        return date.fromisoformat(value)
    except ValueError as exc:
        raise argparse.ArgumentTypeError("日期必须使用 YYYY-MM-DD 格式。") from exc


def build_parser() -> argparse.ArgumentParser:
    """创建复核扫描命令行参数。"""

    parser = argparse.ArgumentParser(
        description="报告待复核或已超过复核日期的错误经验，不修改错题本。"
    )
    parser.add_argument(
        "--notebook",
        type=Path,
        default=DEFAULT_NOTEBOOK,
        help="错题本 Markdown 路径",
    )
    parser.add_argument(
        "--as-of",
        type=parse_as_of,
        default=date.today(),
        help="扫描基准日期，格式 YYYY-MM-DD，默认今天",
    )
    parser.add_argument(
        "--max-results",
        type=int,
        default=20,
        help="最多显示的待处理条目数，默认 20",
    )
    parser.add_argument(
        "--fail-on-action",
        action="store_true",
        help="存在待复核或已到期条目时返回退出码 1",
    )
    return parser


def main() -> int:
    """输出生命周期汇总和需要处理的条目。"""

    args = build_parser().parse_args()
    if args.max_results < 1:
        print("错误：--max-results 必须大于 0。", file=sys.stderr)
        return 2
    if not args.notebook.is_file():
        print(f"错误：错题本不存在：{args.notebook}", file=sys.stderr)
        return 2

    text = args.notebook.read_text(encoding="utf-8")
    _, _, entries = parse_notebook(text)
    labels = [(lifecycle_label(entry.lines, args.as_of), entry) for entry in entries]
    counts = Counter(label for label, _ in labels)
    actionable = [item for item in labels if item[0] in ("待复核", "已到期")]

    print(f"# 错题本复核扫描（截至 {args.as_of.isoformat()}）")
    print()
    print(f"- 总条目：{len(entries)}")
    for label in ("有效", "待复核", "已到期", "已失效", "已替代"):
        print(f"- {label}：{counts[label]}")

    print("\n## 待处理条目\n")
    if not actionable:
        print("没有待复核或已到期条目。")
        return 0

    for label, entry in actionable[: args.max_results]:
        values = extract_lifecycle(entry.lines)
        print(
            f"- [{label}] {entry.category} / {entry.title}；"
            f"最后验证：{values['last_verified']}；下次复核：{values['review_by']}"
        )
    remaining = len(actionable) - args.max_results
    if remaining > 0:
        print(f"- 其余 {remaining} 条未展开；使用 --max-results 调整显示数量。")

    return 1 if args.fail_on_action else 0


if __name__ == "__main__":
    raise SystemExit(main())
